src/indexing.jl

```diff
@@ -176,9 +210,54 @@ Constructs a index for all the files located within  `dir`.
 Same as [`index_checkpoint_files`] except not restricted to files created by Checkpoints.jl.
 """
 function index_files(dir::AbstractPath)
+    depnames_lookup = _depnames_lookup()
     map(Iterators.filter(isfile, walkpath(dir))) do path
-        return IndexEntry(path, dir)
+        return IndexEntry(path, dir, depnames_lookup)
     end
 end
 
 index_files(dir) = index_files(Path(dir))
+
+"""
+    search(name::AbstractString, index)
+
+Returns elements where `name` matches either the full checkpoint name or deprecated names.
+If the `name` is deprecated then a deprecation warning is thrown.
+
+# Arguments
+- `name`: The full checkpoint name to search for (ie: `"Forecasters.forecasts"`)
+- `index`: Iterable of `IndexEntry` elements
+"""
+function search(name::AbstractString, index)
+    results = filter(index) do idx
+        (
+            checkpoint_fullname(idx) == name ||
+            name in idx.deprecated_names
+        )
+    end
+
+    isempty(results) && return results
+
+    fullname = checkpoint_fullname(first(results))
+    name == fullname || Base.depwarn("$name has been deprecated to $fullname.", :search)
+    return results
+end
+
+
+# Utility function for generating a checkpoint name lookup table from the current registry
+function _depnames_lookup()
```

Comment:
The amount of added complexity here makes me wonder if we're accounting for deprecations in the correct way. I'm not sure we should allow a deprecation to point to another deprecation.